modules/mining-pipeline/github-project-miner/src/main/kotlin/org/cafejojo/schaapi/miningpipeline/miner/github/VersionVerifier.kt

```
@@ -0,0 +1,53 @@
+package org.cafejojo.schaapi.miningpipeline.miner.github
+
+import mu.KLogging
+import org.apache.maven.shared.invoker.DefaultInvocationRequest
+import org.apache.maven.shared.invoker.DefaultInvoker
+import org.cafejojo.schaapi.models.project.JavaMavenProject
+import java.io.File
+
+/**
+ * Verifies (dynamically) if projects are using the correct library version.
+ *
+ * @param groupId group id of library maven project should depend on
+ * @param artifactId artifact id of library maven project should depend on
+ * @param version version of library maven project should depend on
+ * @property displayOutput true iff output should be logged at INFO level
+ */
+class VersionVerifier(groupId: String, artifactId: String, version: String, private val displayOutput: Boolean = true) {
+    private companion object : KLogging()
+
+    private val query = "$groupId:$artifactId:jar:$version"
+
+    /**
+     * Verifies (dynamically) if the given [project] is using the correct library version.
+     *
+     * @param project a user project of the library
+     */
+    fun verify(project: JavaMavenProject) =
+        if (createMavenInvoker(project).execute(createMavenInvocationRequest(project)).exitCode != 0) false
+        else getDependencies(project).any { it.startsWith(query) }
```

So... how does it work? I don't understand what's going on.

---

First maven is invoked. If this was unsuccessful, we return false. Then, using `getDependencies` we retrieve the file containing the list of dependencies and parse it to a list. Finally, we 'query' this list for entries starting with the query defined on line 20.

---

Shouldn't we throw an exception if Maven could not be invoked?

---

I agree, the pipeline should break or at least log something if we cannot find a Maven install and not just fail silently.

---

The (original) reasoning is: If a version cannot be verified, we consider this version invalid/wrong. This will have no impact since there are a) enough projects anyway, b) we will fail on this same thing in a later stage. To make sure there's no effect on the whole process, no exceptions are thrown.

I can imagine that there might be unexpected results when projects suddenly get filtered out. I see why a log message might then be useful. However, that same reasoning also applies when maven does work but the outcome is simply 'false'.

So: we can indeed add a log message in the case maven fails, however in my mind this would be kind of inconsistent.

I really get (and agree with) the it's not a `MavenVerifier` reasoning. However the issue here is that that part is verified (at the moment) later in the pipeline, but we already need this information.

---

Ah ok, I see your point. In that case, I agree that it is sufficient for this to fail at a later stage.

---

Do you think it would be appropriate to rename the class to `MavenVersionVerifier`, to distinguish it from other, hypothetical `VersionVerifier`s that do not use Maven?

---

Hm to me that sounds like it verifies the maven version rather than the library version

---

`MavenLibraryVersionVerifier`?